### ISSUE TITLE
question: useful to include automatic WCAG contrast ratio check (js script)?

### ISSUE THREAD

**Issue by brianhewitt** [kss-node/kss-node#100]
Would anyone find it useful to have an optional script to calculate WCAG contrast ratios on colors? We have a color section in our internal style guide, and I've added a script to calculate and note the contrast ratios in the color section.

If other people would find that useful, I could convert that to something more modular (probably with no jquery dependency). 

![screen shot 2014-07-22 at 10 20 00 am](https://cloud.githubusercontent.com/assets/3649834/3800111/ba114fee-1bf4-11e4-85f7-b4952c1a6ff5.jpg)

**Comment by JohnAlbin** [comment]
I'd love to see this! Is there a publicly available example?

**Comment by brianhewitt** [comment]
Didn't mean to ignore this. I'll try to dig up the code and/or an example.

**Comment by ryuran** [comment]
Hello !
I think I can do it in addition to my `Colors` block used here http://developers.20minutes.fr/colette/section-design.html#kssref-design-colors

I think to add this `Colors` block into kss-node

**Comment by brianhewitt** [comment]
Recent work on this kind of thing reminded me about my request for this here, lo those many years ago, so I thought I'd revisit this in case anyone is interested.

My [initial version](https://github.com/brianhewitt/kss-wcag-contrast-calculator/blob/master/kss-contrast-calc.js) of this used custom JS for luminance calculation and comparisons, in part cobbled together from other sources.

Lately, however, I've been using the very handy color manipulation and calculation tools in [TinyColor](https://bgrins.github.io/TinyColor/), which very conveniently provides built in WCAG compliance level and text size properties to pass to an `isReadable` method. Here's how I've got this set up to parse colors and write the results (in a custom template). 

`'use strict';

const colorBlocks = document.querySelectorAll('.sg__color-sample');

let colorStatusClass = (color, contrastColor) => { // check readability and return appropriate classes

  let colorStatusIndicators = {
    colorStatusClass: '',
    statusIconClass: ''
  };

  if ( tinycolor.isReadable(color, contrastColor, {level:'AA',size:'small'}) ) {
    colorStatusIndicators.colorStatusClass = 'wcag-aa-pass';
    colorStatusIndicators.statusIconClass = 'fa-star';
  }

  else if (tinycolor.isReadable(color, contrastColor, {level:'AA',size:'large'})  ) {
    colorStatusIndicators.colorStatusClass = 'wcag-a-pass';
    colorStatusIndicators.statusIconClass = 'fa-star-half-full';
  }

  else {
    colorStatusIndicators.colorStatusClass = 'wcag-a-fail';
    colorStatusIndicators.statusIconClass = 'fa-ban';
  }

  return colorStatusIndicators;
};

if (colorBlocks.length) {
  const whiteColor = tinycolor('fff');

  // containers to write contrast status into
  const wcagStatusDiv_white = '<span class="kss-color-status kss-color-status--white"></span>';
  const wcagStatusDiv_text = '<span class="kss-color-status kss-color-status--text"></span>';

  window.addEventListener('load', () =>{
    const blackBlock = document.querySelectorAll('.black')[0]; // specify text/darkest color sample class
    const blackColor = tinycolor(window.getComputedStyle(blackBlock).getPropertyValue('background-color'));

    /* loop through color sample blocks to
    get and set contrast ratios and WCAG AA pass/fail classes
     */
    colorBlocks.forEach( (colorItem) => {
      let colorBlockParentElem = colorItem.parentElement;

      // get color
      let thisColorBlockBg = window.getComputedStyle(colorItem).getPropertyValue('background-color');
      let thisColorBlockColor = tinycolor(thisColorBlockBg);

      let colorValElem = colorBlockParentElem.querySelectorAll('.sg__color-value')[0];

      if (colorValElem) {
        colorValElem.innerHTML = thisColorBlockColor.toString('rgb') + '<br>' + thisColorBlockColor.toString('hex6') + '<br>' + thisColorBlockColor.toString('hsv');
      }

      // WCAG AA ratio numbers
      let colorRatio_white = Math.round(tinycolor.readability(whiteColor, thisColorBlockColor) * 10) / 10;
      let colorRatio_black = Math.round(tinycolor.readability(blackColor, thisColorBlockColor) * 10) / 10;

      let wcagContrast_whiteColor_Container = colorBlockParentElem.querySelectorAll('.kss-wcag__contrast--white')[0];
      let wcagContrast_blackColor_Container = colorBlockParentElem.querySelectorAll('.kss-wcag__contrast--text')[0]; // black(ish)

      if (wcagContrast_whiteColor_Container) { // display of white contrast ratio info
        // get elements
        let ratioElem_white = wcagContrast_whiteColor_Container.querySelectorAll('.kss-wcag__ratio')[0];
        let colorStatusElem_white = wcagContrast_whiteColor_Container.querySelectorAll('.kss-color-status--white')[0];
        let colorStatusIcon_white = colorStatusElem_white.firstChild;

        // write color ratio to style guide
        ratioElem_white.textContent = colorRatio_white;

        // get appropriate WCAG AA pass/fail classes
        let colorStatusIndicators_white = colorStatusClass(thisColorBlockColor, whiteColor);

        // set WCAG AA pass/fail classes
        colorStatusElem_white.classList.add(colorStatusIndicators_white.colorStatusClass);
        colorStatusIcon_white.classList.add(colorStatusIndicators_white.statusIconClass);
      }

      if (wcagContrast_blackColor_Container) { // display of text color (black) contrast ratio info
        // get elements
        let ratioElem_black = wcagContrast_blackColor_Container.querySelectorAll('.kss-wcag__ratio')[0];
        let colorStatusElem_black = wcagContrast_blackColor_Container.querySelectorAll('.kss-color-status--text')[0];
        let colorStatusIcon_black = colorStatusElem_black.firstChild;

        // write color ratio to style guide
        ratioElem_black.textContent = colorRatio_black;

        // get appropriate WCAG AA pass/fail classes
        let colorStatusIndicators_black = colorStatusClass(thisColorBlockColor, blackColor);

        // set WCAG AA pass/fail classes
        colorStatusElem_black.classList.add(colorStatusIndicators_black.colorStatusClass);
        colorStatusIcon_black.classList.add(colorStatusIndicators_black.statusIconClass);
      }
    });
  });
}`

At some point, I'd actually to automatically select the darkest color from the lowest calculated luminance, rather than manually specifying a the class of the element that's showing that color.
